From 17913d8783084b3587d68a9451237008b18b7067 Mon Sep 17 00:00:00 2001
From: Georg Richter <georg@mariadb.com>
Date: Tue, 27 Oct 2026 16:26:44 +0100
Subject: [PATCH 4/5] Branch 3.4 after 3.4.11: Add cmake option
 RUN_UNSTABLE_TESTS for unit tests

Commit 539c7c275ff8e0410a4996745d51e2ddee8e8bba of the 3.4 branch of
https://github.com/mariadb-corporation/mariadb-connector-c, as it is.

Add cmake option RUN_UNSTABLE_TESTS for unit tests

Add the RUN_UNSTABLE_TESTS CMake option (default OFF) to exclude flaky
or timing-sensitive tests from default builds and CI runs.

Mark test_tls_openssl_err2 as unstable because it frequently times out
on GitHub Actions. Manual verification confirmed that OpenSSL correctly
returns x509_error=2 for the certificate provided by the fake server.
---
 CMakeLists.txt                     | 2 ++
 unittest/libmariadb/CMakeLists.txt | 5 ++++-
 unittest/libmariadb/tls.c.in       | 8 ++++++--
 3 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index d4c6280..e4e56bf 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -37,6 +37,8 @@ ADD_OPTION(WITH_TOOLS
            "Build tools"
            OFF)
 
+ADD_OPTION(RUN_UNSTABLE_TESTS "Run unstable tests" OFF)
+
 IF(WITH_BOOST_CONTEXT)
   PROJECT(mariadb-connector-c LANGUAGES C CXX)
 ELSE()
diff --git a/unittest/libmariadb/CMakeLists.txt b/unittest/libmariadb/CMakeLists.txt
index 0b5ebc8..a69418a 100644
--- a/unittest/libmariadb/CMakeLists.txt
+++ b/unittest/libmariadb/CMakeLists.txt
@@ -17,6 +17,10 @@ IF(SKIP_TESTS)
   RETURN()
 ENDIF()
 
+IF(RUN_UNSTABLE_TESTS)
+  ADD_DEFINITIONS(-DRUN_UNSTABLE_TESTS)
+ENDIF()
+
 ENABLE_TESTING()
 
 IF(WIN32 AND DEFINED ENV{TRAVIS})
@@ -106,4 +110,3 @@ FOREACH(FILE ${ADDITIONAL_FILES})
                  ${CC_BINARY_DIR}/unittest/libmariadb/${FILE}
                  COPYONLY)
 ENDFOREACH()
-
diff --git a/unittest/libmariadb/tls.c.in b/unittest/libmariadb/tls.c.in
index 9d54dd0..3ed232a 100644
--- a/unittest/libmariadb/tls.c.in
+++ b/unittest/libmariadb/tls.c.in
@@ -1165,11 +1165,15 @@ end:
   return rc;
 }
 
-#ifdef HAVE_OPENSSL
+#if defined(RUN_UNSTABLE_TESTS) && defined(HAVE_OPENSSL)
 /*
   The client trusts the intermediate but not the root. The chain reaches the
   trust store but can't be completed: X509_V_ERR_UNABLE_TO_GET_ISSUER_CERT (2),
   because ctx->num_untrusted < num.
+
+  Please note that this test frequently fails on github.actions with timeout
+  error. To run this test please specify -DRUN_UNSTABLE_TESTS=ON when running
+  cmake
 */
 static int test_tls_openssl_err2(MYSQL *my __attribute__((unused)))
 {
@@ -1233,7 +1237,7 @@ struct my_tests_st my_tests[] = {
 #ifndef HAVE_SCHANNEL
   {"test_tls_verify_leaf_without_ca", test_tls_verify_leaf_without_ca, TEST_CONNECTION_NEW, 0, NULL, NULL},
 #endif
-#ifdef HAVE_OPENSSL
+#if defined(RUN_UNSTABLE_TESTS) && defined(HAVE_OPENSSL)
   {"test_tls_openssl_err2", test_tls_openssl_err2, TEST_CONNECTION_NEW, 0, NULL, NULL},
 #endif
   {"test_tls_verify_critical_ext", test_tls_verify_critical_ext, TEST_CONNECTION_NEW, 0, NULL, NULL},
-- 
2.53.0

